### PR TITLE
Default pipeline - add packing

### LINKED ISSUE #361: Add `pack-matmul` and `pack-vnni` to the default pipeline
When creating the MLP generator (#360), I noticed the matmuls weren't being packed on `tpp-opt --default-tpp-passes`. It seems those passes aren't on our standard pass.

We need to add them to the default pipeline so that we can automatically optimise MLP without the need of running `tpp-opt` twice.

We may need to pass the tile sizes to the pass, so this could be an extra option on `tpp-opt` and `tpp-run`?

@adam-smnk

### PR DESCRIPTION
**Pull request by adam-smnk**
Adds various existing packing passes to the default pipeline.
These passes are run with either their own default blocking factors or values hardcoded for the default pipeline.
Additionally, other support passes are added to improve packing performance such as pack propagation or folding.

The default matmul packing is disabled in some benchmarks in which the pass highly reduces performance.

Fixes #361 
Work towards #369

### CODE REVIEW

**Review by rengolin — COMMENTED** [review]
I'm surprised this didn't change any of the existing tests, especially in the `Models` directory. Perhaps we should convert all integration tests to the default pass sooner than later.

**Comment by adam-smnk** [comment]
> I'm surprised this didn't change any of the existing tests, especially in the `Models` directory. Perhaps we should convert all integration tests to the default pass sooner than later.

Don't worry, some models do fail now. That is why I need to common up dispatch calls before I start updating like 100 of the existing dispatch checks 😜

File: test/Passes/DefaultPipeline/tpp-mapping.mlir [hunk 1]
```diff
@@ -1,13 +1,89 @@
...
 }
 
 // CHECK-LABEL: func.func @conv_to_matmul(
 // CHECK-NOT: linalg.conv_2d_nhwc_hwcf
-// CHECK: linalg.matmul
+// CHECK: scf.for
+// CHECK:   scf.for
+// CHECK:     scf.for
+// CHECK:       scf.for
+// CHECK:         tensor.extract_slice
```
[the remark this answers is not in the document]
Comment (adam-smnk) [reply test/Passes/DefaultPipeline/tpp-mapping.mlir]: Added

**Comment by adam-smnk** [comment]
Default matmul packing currently completely decimates benchmark performance so, it'll be disabled there.
However, `pack-matmul` will remain in the default pipeline as long term we still need this pass. In some cases, packing is necessary to enable other transformations such as VNNI packing. The results of matmul packing will remain tested in various passes tests' IR checks.

File: test/Passes/DefaultPipeline/tpp-mapping.mlir [hunk 2]
```diff
@@ -1,13 +1,90 @@
-// RUN: tpp-opt %s -tpp-mapping -split-input-file | FileCheck %s
+// RUN: tpp-opt %s -tpp-mapping -cleanup -split-input-file | FileCheck %s
```
Comment (rengolin) [inline test/Passes/DefaultPipeline/tpp-mapping.mlir:1]: Is this cleanup just for the test or is it really necessary? If the latter, then it should go in the pipeline after `createGeneralizeTensorPackAndUnPackPass`.

**Review by rengolin — COMMENTED** [review]
We may have to break the `TppMappingPass` later, but it's ok for now.

**Comment by rengolin** [comment]
> Default matmul packing currently completely decimates benchmark performance so, it'll be disabled there. However, `pack-matmul` will remain in the default pipeline as long term we still need this pass. In some cases, packing is necessary to enable other transformations such as VNNI packing. The results of matmul packing will remain tested in various passes tests' IR checks.

It may sound like a silly test, but that's a good canary to show us the side-effects of pack/tile transforms.

File: test/Passes/DefaultPipeline/tpp-mapping.mlir (hunk 2 above)
Comment (adam-smnk) [reply test/Passes/DefaultPipeline/tpp-mapping.mlir:1]: It is already present in the default pipeline as we run cleanup after every major stage i.e., mapping, lowering etc.
It is mainly for the test as it produces a bit nicer IR. But you're right, it should be in the mapping pass to limit number of possible combinations.

**Comment by adam-smnk** [comment]
Canary performance regression due to matmul packing is now captured in #410.
Cause of performance regression in the bf16 MLP models is unknown. However, the compilation pipeline cannot be currently optimized for them due to the runtime issue #409.

**Review by rengolin — APPROVED** [review]
There are some performance regressions but we need to merge this before the dungeon. I'll approve this so we can do that, but we need to make sure we come back from the crash in performance by spotting and fixing all of the issues in the default pipeline (not manual runs, special flags, etc).

**Pull request merged**